src/app/rmm/me.ts

```diff
@@ -0,0 +1,52 @@
+// --------- BEGIN RUNBOX LICENSE ---------
+// Copyright (C) 2016-2018 Runbox Solutions AS (runbox.com).
+// 
+// This file is part of Runbox 7.
+// 
+// Runbox 7 is free software: You can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the
+// Free Software Foundation, either version 3 of the License, or (at your
+// option) any later version.
+// 
+// Runbox 7 is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+// 
+// You should have received a copy of the GNU General Public License
+// along with Runbox 7. If not, see <https://www.gnu.org/licenses/>.
+// ---------- END RUNBOX LICENSE ----------
+import { timeout, share } from 'rxjs/operators';
+import { RMM } from '../rmm';
+
+export class Me {
```

Comment:
This goes for all the new RMM classes (which are a good idea: the one giant rmmapi has gotten a bit silly by now...), but I think it'd be good to make them `@Injectable` so that every component using them doesn't need to create them by itself (which also makes testing near impossible).

Right now every component using `RMM::Me` has to created it by itself and then load the values (which I'd also rather have eliminated, as described in the comment above). That means that we're performing a query to `/rest/v1/me` for every component using `RMM::Me`, which is pretty wasteful.

Making this `@Injectable` means that any object using it can just declare a depedency on this in its constructor arguments (like we do now with rmmapi) and then only one copy is being created in the entire app, and no unnecessary queries are performed. Also, it becomes possible to replace them with stubs and mocks in tests.

Reply:
definately should not call multiple times... the idea is that 'rmm' is 1 instance.

So this.rmm.aliases.load() will probably consume aliases endpoint. and store data this.rmm.aliases.data so other classes can use this.rmm.aliases.data without reloading it.

Reply:
Also,

Dont add @Injectable in RMM Class..

First im trying to avoid all that @*unecessary-stuff*

The RMM class is injectable, and 'this.rmm' is a normal OO class that has atributes, instances, normal stuff.

If you see the Me load() method, it has a 

`
     load() {
         if ( this.data ) { return; }
`

So 3 places in profiles call load() at the same time, and neither of the calls were terminated. So it calls 3 times. If data were to have a value, it would not call the endpoint again.

This class has an attribute which is this.is_busy which is true when its fetching values. To fix the multiple calls it possible to alter the code to:
`
     load() {
         if ( this.data || this.is_busy ) { return; }
`

That will make the endpoint be called only 1 time.

We can of course change how this works but lets keep it simple OO.

Reply:
Imho `@Injectable` *is* simple OO. It's just the readymade implementation of one of the SOLID rules, which is arguably basic OO :)

I see the point about the single instance: I didn't notice that Me etc are all part of RMM and you're supposed to use it as a whole. In that case you will indeed have only one copy of everything.

But it does feel, currently, like a bit of a downgrade from RunboxMe: it requires me to pull in the entire God-object as a dependency effectively, and an object that does its own UI for error handling. I like the idea of keeping things simple and synchronous as much as possible, since it does make everything easier, but as it stands I think things are made a bit *too* simple. If you want to show some data in the template then cool: it works for that, if you just bind to `Me.data`. But consider `StorageService` for instance: it needs to know part of the Me, the uid, to implement user-specific localStorage. There's no template, nothing to bind to. It needs a way to say "give me the UID, once it becomes available, and then lookup the localStorage with the key mashed together with the UID". The `RMM.Me` class doesn't make it possible to implement that. Grabbing `.data` is not good enough: it may be not there yet. I can call `.load()` and subscribe on the result... but if someone called .load before me then I'm either duplicating calls (currently), or if we apply the `if ( this.data || this.is_busy ) { return }` fix you mentioned, I'm not going to get the HTTP query object and will have nothing to subscribe to.

We keep can keep as it is, sure. But I don't see myself using it at this point just yet :)

Reply:
I did not consider any localstorage for that RMM class.. but the data is in this.rmm.me.data so it could be easily jsonstringified or stored in localstorage when its fetched, or processed for xapian when its fetched.

If you think its too simple, that is good, that was the point :)

Any of it can be changed, upgraded, etc.. im just trying to create this 'model' so its easy to create screens that list data and allow to edit them via simple crud.

When this version is better you may use it

Reply:
It's not about localstorage itself, it's about the asynchronous nature of the data :) Your approach works fine if the async data from your components is put straight into the Angular template, but is unfit for cases if you need some extra processing done to it.

> If you think its too simple, that is good, that was the point :)

Aha, I think I see your approach now :) It definitely does work well for simple cases, with inbuilt error handling and all that – makes sense. For the heavy lifting we can still use the async stuff where necessary.

Reply:
One nitpick that I have though: can we get some types added into this and make this consistent?

`load()` sometimes returns the async query that it's running, which will resolve to the data later, but sometimes it does not. It may lead to a situation where a code works one day, because it was the first component to call .load(), and a week later some other component gets to it first and the old one will end up doing `undefined.subscribe` because of reasons beyond its control.

I think it'd be safer to have the type declared in Typescript: make it `load(): void { }` or `load(): Observable<any> { }`, which ever you see more fitting (the former is probably easier, the latter may be more useful), and this will protect us from accidentally misusing it.

Reply:
if undefined.subscribe situation exists this will result in error and thats not wanted.

I will use the Observable<any> suggestion

Reply:
You can probably use `Observable.of` (I'm not sure if that's what it's called) to return something compatible in case the data is already there (the early exit from `load()`)